File: okmock-interceptor/src/main/java/nolambda/gadb/okmock/OkMock.kt
package nolambda.gadb.okmock

import nolambda.gadb.okmock.adapter.DefaultSerializer
import nolambda.gadb.okmock.adapter.OkMockAdapter
import nolambda.gadb.okmock.adapter.OkMockPayload
import nolambda.gadb.okmock.adapter.PayloadParser
import nolambda.gadb.okmock.server.OkMockServer
import nolambda.gadb.okmock.server.OkMockServerImpl
import okhttp3.Interceptor
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.Protocol
import okhttp3.Request
import okhttp3.Response
import okhttp3.ResponseBody.Companion.toResponseBody

class OkMock(
    okMockServer: OkMockServer = OkMockServerImpl(),
    adapter: OkMockAdapter = OkMockAdapter(
        parser = PayloadParser(),
        serializer = DefaultSerializer()
    )
) : Interceptor {

    private val sender = { req: Request, payload: OkMockPayload ->
        okMockServer.send(adapter.serializer.serialize(req, payload))
    }

    init {
        okMockServer.start()
        okMockServer.listen {
            registerMockResponse(adapter.parser.parse(it))
        }
    }

    private val mockResponses = mutableListOf<OkMockPayload>()

    private fun registerMockResponse(payloads: List<OkMockPayload>) {
        mockResponses.clear()
        mockResponses.addAll(payloads)
    }

    override fun intercept(chain: Interceptor.Chain): Response {
        val request = chain.request()
        val (payload, mockResponse) = getMockResponse(request) ?: return chain.proceed(request)

        sender(request, payload)

        return mockResponse
    }

    private fun getMockOrNull(info: PartialRequestInfo): OkMockPayload? {
        val haveMethod = mockResponses.any { it.method == info.method }
        if (haveMethod.not()) return null

        val url = info.url
        return mockResponses.find { it.path.matches(url) }
    }

    private fun getMockResponse(request: Request): Pair<OkMockPayload, Response>? {
        val url: String = request.url.toString()
        val method: String = request.method
        val partialRequest = PartialRequestInfo(url, method)
        val payload = getMockOrNull(partialRequest) ?: return null
        val responseBody = payload.body.toResponseBody("application/text".toMediaTypeOrNull())

        val builder: Response.Builder = Response.Builder()
            .request(request)
            .protocol(Protocol.HTTP_1_1)
            .code(payload.code)
            .message(payload.message)
            .receivedResponseAtMillis(System.currentTimeMillis())
            .body(responseBody)

        payload.headers.forEach {
            builder.header(it.key, it.value)
        }
        return payload to builder.build()
    }
}

File: flutter_okmock/sample/lib/OkMock.dart
import 'package:dio/dio.dart';
import 'package:sample/OkMockAdapter.dart';
import 'package:sample/OkMockPayload.dart';
import 'package:sample/OkMockServer.dart';
import 'package:sample/PartialRequestInfo.dart';

class OkMock extends Interceptor {
  static const String CHANNEL_MOCK = "mock";
  static const String CHANNEL_CLEAR = "clear";

  Dio dio;
  OkMockServer server;
  OkMockAdapter adapter;

  List<OkMockPayload> mockResponses = new List();

  OkMock({Dio dio, OkMockServer server, OkMockAdapter adapter}) {
    _init(dio, server, adapter);
  }

  OkMock.createDefault(Dio dio) {
    _init(
        dio,
        OkMockServerImpl(),
        OkMockAdapter(
            deserializer: PayloadDeserializer(),
            serializer: DefaultSerializer()));
  }

  void _init(Dio dio, OkMockServerImpl server, OkMockAdapter adapter) {
    this.dio = dio;
    this.server = server;
    this.adapter = adapter;

    server.listen(CHANNEL_MOCK, (mock) {
      _registerMockResponse(adapter.deserializer.deserialize(mock));
    });
    server.listen(CHANNEL_CLEAR, (_) => _clearMockResponse());
    server.start();
  }

  void _registerMockResponse(List<OkMockPayload> payloads) {
    mockResponses.clear();
    mockResponses.addAll(payloads);
  }

  void _clearMockResponse() {
    print("Clearing mocks…");
    mockResponses.clear();
  }

  @override
  Future onRequest(RequestOptions options) async {
    OkMockPayload payload = await getMockResponse(options);
    if (payload != null) {
      server.send(adapter.serializer.serialize(options, payload));

      return dio.resolve(payload.body);
    }
    return options;
  }

  Future<OkMockPayload> getMockResponse(RequestOptions options) async {
    if (mockResponses.isEmpty) return null;

    String method = options.method;
    String url = options.uri.toString();

    print("=> method $method url: $url");

    PartialRequestInfo info = PartialRequestInfo(method: method, url: url);
    return mockResponses.singleWhere(
        (element) => element.method == info.method && element.path == info.url);
  }
}
